### PR TITLE
Understanding concurrency in Go

### LINKED ISSUE #3397: Understanding Concurrency in GoLang
### Introduction paragraph (2-3 paragraphs):
Smaller sub-programs make up large programs. What is ideal is for these smaller sub-programs to be able to run at the same time. This is what is referred to as Concurrency. 
Concurrency refers to a program's ability to be divided into independent tasks that can be executed at the same time.
This simply means a program that has two or more tasks that can execute individually of each other but are still part of the same program.

Golang has in-built concurrency using goroutines and channels. Concurrency in Golang is also simple and easy to use. 
In this tutorial, we will be using goroutines to run concurrency within a Go application. 

### Key takeaways: 
1. Golang concurrency basics using goroutines. 
2. Difference between concurrency and parallel programming.
3. How to write your program as a collection of independent processes.

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Anitaachu**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #3397.

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by Anitaachu** [comment]
This closes #3397

**Comment by ahmadmardeni1** [comment]
Good afternoon and thank you for submitting your article for publishing. @Anitaachu 
Your article has been entered into our queue and should be reviewed (and published pending any changes) as soon as a peer reviewer is finished with the ones in the queue before it. :rocket:
Thank you.

**Comment by hectorkambow** [comment]
@Anitaachu 
There are TWO index files in this PR. 

🤔

**Comment by hectorkambow** [comment]
[Understanding concurrency in Go #3666.pdf](https://github.com/section-engineering-education/engineering-education/files/7117954/Understanding.concurrency.in.Go.3666.pdf)

plagirism check done 

@Anitaachu 

It will improve the article to go through the red flags in the report.

**Comment by Anitaachu** [comment]
Requested changes have been and I ran the test again

**Comment by adrianmurage** [comment]
@Anitaachu You seem to have added images and an article related to "integrating-graphql-api-in-a-django-application"

Kindly remove any files unrelated to "Understanding concurrency in Go"

**Review by adrianmurage — CHANGES_REQUESTED** [review]
@Anitaachu 

I like the article thus far but It feels incomplete. 
As a reader, I don't really get to see the effects of the implementation of concurrency. You introduced the aspect of time before implementing concurrency but never revisit it once you implemented concurrency you don't revisit the time concept. 

It would be really great if you could clearly illustrate how introducing concurrency affects the performance of the application. Otherwise, the significance of concurrency is lost on the reader. 

Looking forward to seeing how you do this. Kindly @ me when the necessary changes are made.

**Comment by adrianmurage** [comment]
@Anitaachu ?

**Comment by Anitaachu** [comment]
@adrianmurage I'm sorry for the delayed response. I noticed some errors and had to restructure the drafts. All requested changes have been made. Thank you

File: articles/concurrency-in-go/index.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+In programming, concurrency is "*the composition of independently executed processes...*". Concurrency refers to the ability of a program to divide into independent tasks that can all execute at the same time. it could be handling many requests from users or sending many responses to users.
+
+### **Concurrency vs Parallelism**
+
+A lot of developers intertwine concurrency with parallelism. Though these two terms may seem similar and are often confusing and substituted for the other, they do not mean the same thing. Concurrency is the ability to deal with lots of programs at the same time. These programs execute at the same time one after another. For example, a coffee shop with many customers requesting a cup of their choice of coffee. As a customer gets served, the order of another customer processes. Every customer gets served from the same coffee machine. Parallelism is the ability to do lots of programs at the same time. The same analogy applies, although, here there are different coffee machines.
+
+In both scenarios, customers are served at the same time. Parallelism is a tool used by concurrency to achieve its goal however, it is not the goal of concurrency.
```
Comment (adrianmurage) [inline articles/concurrency-in-go/index.md:30]:
```suggestion
In both scenarios, customers are served at the same time. Parallelism is a tool used by concurrency to achieve its goal, however, it is not the goal of concurrency.

```

File: articles/concurrency-in-go/index.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+ [Goroutines](https://www.golangprograms.com/goroutines.html) are functions that run at the same time alongside other code or programs. Goroutines are unique to Go. They’re not OS threads though they may be considered lightweight threads. Goroutines are deeply integrated with Go runtime. Go's runtime monitors goroutine behavior and suspends them when they become blocked, then resumes them once they are not blocked. This makes them preemptible in some ways, but only when the goroutine is blocking. The runtime and the logic of a goroutine work together. Goroutines can communicate with one another and synchronize their execution. In Go, one of the synchronization elements is called a channel. Though they are used to synchronize memory access, they are best used for communicating information between goroutines. When data is to be shared across goroutines while performing a concurrent activity as a goroutine, channels function as a conduit (pipe) between the goroutines and provide a method that ensures asynchronous exchange.
```
Comment (adrianmurage) [inline articles/concurrency-in-go/index.md:36]:
```suggestion
 [Goroutines](https://www.golangprograms.com/goroutines.html) are functions, unique to Go, that run at the same time alongside other code or programs.

They’re not OS threads, though, they may be considered lightweight threads.

Goroutines are deeply integrated with Go's runtime. Go's runtime monitors goroutine behavior and suspends them when they become blocked, then resumes them once they are unblocked. 

This makes them preemptible in some ways, but only when the goroutine is blocking.

The runtime and the logic of a goroutine work together. Goroutines can communicate with one another and synchronize their execution. In Go, one of the synchronization elements is called a channel. 

Though channels are used to synchronize memory access, they are best used for communicating information between goroutines. When data is to be shared across goroutines while performing a concurrent activity as a goroutine, channels function as a conduit (pipe) between the goroutines and provide a method that ensures asynchronous exchange.

```

File: articles/concurrency-in-go/index.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+This program contains two functions: the *request function* and the *main function*. The request function sends an *HTTP* request and also returns a response or an error if any. 
+
+The *main function* is set to hold more than two arguments. Next, we iterate over the URLs and send requests.
+
+In our terminal, we will pass in some URLs as a request and see how much time it took for our request to complete. Now, using the *time* command in your bash terminal pass in these URLs: `time go run main.go google.com github.com youtube.com amazon.com plaid.com gitlab.com heroku.com notion.com` and run the command. 
+
+![Zero concurrency](/engineering-education/concurrency-in-go/noconcurrency.png)
+
+The program was executed at exactly, 8.937s
```
Comment (adrianmurage) [inline articles/concurrency-in-go/index.md:79]: This time does not reflect the time shown in your image above

**Review by adrianmurage — DISMISSED** [review]
@Anitaachu  Awesome job with the update. You implemented the recommendations in a satisfactory manner. Thank you!

I've pointed out a few more changes in the formating that you need to address after which we can mark the article ready for final review

**Comment by Anitaachu** [comment]
Thank you @adrianmurage. All requested changes and corrections have been made.

**Comment by adrianmurage** [comment]
@Anitaachu 
You still have files related to the `integrating-graphql-api-in-a-django-application` article in your PR.
Kindly delete them.

**Comment by Anitaachu** [comment]
@adrianmurage they've been deleted. Thank you.

**Comment by adrianmurage** [comment]
@Anitaachu 
Great job with the article. Kindly go through the changes I made.
Looking forward to seeing more of your work!

**Comment by LinusMuema** [comment]
Hi @Anitaachu 
Awesome job on the article. The final review is done :rocket: 

cc: @adrianmurage you can check my commit for the changes I have made. The image extensions were in caps but it was a small change.

**Pull request merged**